### PR TITLE
owner: Add error check in NewSession

### PR DESCRIPTION
**Pull request by zimulala**
Sometimes the etcd client is closed but context doesn't be close. 
I think make them close together is better. I will do it later.

### CODE REVIEW

**Comment by shenli** [comment]
@zimulala Please fix the CI.

File: owner/manager.go [hunk 1]
```diff
@@ -277,11 +288,17 @@ func init() {
...
-func isContextDone(ctx goctx.Context) bool {
+func contextDone(ctx goctx.Context, err error) error {
 	select {
 	case <-ctx.Done():
-		return true
+		return errors.Trace(ctx.Err())
 	default:
 	}
-	return false
+	if terror.ErrorEqual(err, goctx.Canceled) ||
```
Comment (shenli) [inline owner/manager.go]: Add comment to describe why we should ignore these errors.

**Comment by shenli** [comment]
@coocood @winkyao PTAL

**Comment by winkyao** [comment]
@zimulala Please fix the CI.

File: owner/manager.go [hunk 2]
```diff
@@ -277,11 +288,20 @@ func init() {
 	}
 }
 
-func isContextDone(ctx goctx.Context) bool {
+func contextDone(ctx goctx.Context, err error) error {
```
Comment (winkyao) [inline owner/manager.go]: prefer let the contextDone figure out whether context is done. and add a function like ErrorOccured to check the error, one function is better to do just one thing.

File: owner/fail_test.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	_, err = NewSession(goctx.Background(), "fail_new_serssion", cli, retryCnt, ManagerSessionTTL)
+	isContextDone := terror.ErrorEqual(grpc.ErrClientConnClosing, err) || terror.ErrorEqual(goctx.Canceled, err)
+	c.Assert(isContextDone, IsTrue, Commentf("err %v", err))
+
+	cli, err = clientv3.New(clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: dialTimeout,
+	})
+	gofail.Enable("github.com/pingcap/tidb/owner/closeGrpc", `return(true)`)
+	_, err = NewSession(goctx.Background(), "fail_new_serssion", cli, retryCnt, ManagerSessionTTL)
```
Comment (winkyao) [inline owner/fail_test.go]: miss `isContextDone = terror.ErrorEqual(grpc.ErrClientConnClosing, err) || terror.ErrorEqual(goctx.Canceled, err)`

File: owner/manager.go (hunk 2 above)
Comment (zimulala) [reply owner/manager.go]: This function checks whether the context is done， and we treat these error as the context is done. So I put these logical in one function.

File: owner/fail_test.go (hunk 3 above)
Comment (zimulala) [reply owner/fail_test.go]: It has done on line 75.

**Comment by zimulala** [comment]
PTAL @winkyao @shenli

File: owner/fail_test.go (hunk 3 above)
Comment (shenli) [reply owner/fail_test.go]: It is already checked in line 76, why we should check it in line 84?

Comment (zimulala) [reply owner/fail_test.go]: Done.

**Comment by shenli** [comment]
LGTM

File: owner/manager.go [hunk 4]
```diff
@@ -277,11 +288,20 @@ func init() {
...
 	select {
 	case <-ctx.Done():
-		return true
+		return errors.Trace(ctx.Err())
 	default:
 	}
-	return false
+	// Sometime the ctx isn't closed, but the etcd client is closed,
+	// we need to treat it as if context is done.
+	// TODO: Make sure ctx is closed with etcd client.
```
Comment (coocood) [inline owner/manager.go]: We don't have the cancel function, how can we make sure `ctx` is closed?

File: owner/fail_test.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+func (s *testSuite) TestFailNewSession(c *C) {
+	defer testleak.AfterTest(c)()
+
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: dialTimeout,
+	})
+	gofail.Enable("github.com/pingcap/tidb/owner/closeClient", `return(true)`)
+	_, err = NewSession(goctx.Background(), "fail_new_serssion", cli, retryCnt, ManagerSessionTTL)
+	isContextDone := terror.ErrorEqual(grpc.ErrClientConnClosing, err) || terror.ErrorEqual(goctx.Canceled, err)
```
Comment (coocood) [inline owner/fail_test.go]: Does the returned error changes on each run?

File: owner/manager.go (hunk 4 above)
Comment (zimulala) [reply owner/manager.go]: So I will do it in the future.

**Comment by coocood** [comment]
LGTM

**Pull request merged**